Issue 1989

Naaaaiix opened the pull request:
*** NEW DESCRIPTION (13/10/21) ***

This branch is ready to be reviewed and merged to master. Please, report all bugs you find, but just going to fix the ones related to the implementation itself; if they are minor or side-effects will be derived to new issues to avoid extending the merge.

This branch aims to implement what has been described on the bottom of the description and on #2212:
1. Abstract Responsive bar as a new component: ResponsiveTabsControl (more info on #2212 )
2. Abstract 2 components from the old BackgroundControl: BlockBackgroundControl and BackgroundControl (more info on #2212 )
3. **Implement responsive for background** components (BlockBackgroundControl && BackgroundControl) and the helper of Style API
4. **Adapt Button Maxi, Hover Effects, Icon colour settings, Arrow and Parallax** to new background components and responsive support
5. **Create and update tests** related to the implementation

Notes:
1. Code hasn't been optimized. It means helpers generate more code than needed which reflects into more CSS code. This needs to be fixed during the future optimization.
2. The new ResponsiveTabsControl need some iterations to find how to deal with inner plugin necessities and to be user friendly. I'll derivate a new issue to talk about it.
3. For arrows, the workflow doesn't seem much user friendly. For the moment is getting the last colour layer, but can be more user friendly with the future new UI (will be derivate to a new issue):

https://user-images.githubusercontent.com/50477222/137176477-5cab519c-df7f-4702-bb6c-128fc60cd8d3.mp4

4. As @kyrapieterse commented on chat, the parallax workflow feels strange, as seems you lose part of what has been done (will be derivate to a new issue):

https://user-images.githubusercontent.com/50477222/137177049-678f8266-3a60-4771-87cc-c0ef16d33c3e.mp4

5. Some tests have been disabled; onces merged @Alejandro9kgm will fix them 👍 (thanks buddy!)

---------------------

 **OLD DESCRIPTION**

--------------------

This PR is not for merging directly as is not 100% functional; we might find some bugs and need some changes. Couldn't fix them as we may need some UX implementations. I'll try to make it the most understandable way.

The idea has been to simply implement responsive as a background blocks compatibility. Even after fight against it, I've found some parts they were hard to integrate for the logical of the whole background ecosystem: from the components that creators can use on the settings sidebar, as the scripts in charge of create the HTML and CSS. So, what has been done in this branch:
1. **Implement responsive for background** component (BackgroundControl) and the helper of Style API
2. **Adapt Button Maxi** to new responsive background support
3. **Adapt Arrow components** (ArrowControl & ArrowDisplayer) to new responsive background support
4. **Create and update tests** related to background responsive support

As said before, there are some aspects I would like to share with you as they don't seem much solid for me:
1. Simple background and layers background:
What is what:

| Simple background | Layers background |
| --- | --- |
| <img width="282" alt="Captura de pantalla 2021-10-04 a las 16 26 57" src="https://user-images.githubusercontent.com/50477222/135869277-9b72e477-f9a3-4b92-8eff-f02d9d869f4c.png"> | <img width="282" alt="Captura de pantalla 2021-10-04 a las 16 27 46" src="https://user-images.githubusercontent.com/50477222/135869389-d76e3964-6d7e-4da4-b4c5-c20f9b09b6bb.png"> |
 
Lets imagine a Creator who wants to have a **simple background on `base`** responsive stage, but **goes to `xl` and changes to layers**, and then in **`s` wants to have another simple background**. Yes, is strange, but is possible; and the logics the scripts need to work correctly in that way are complex (they are not working right now, that's a bug you'll find if you test):

https://user-images.githubusercontent.com/50477222/135871661-6ddb8997-8808-4e36-ad20-a2df75e73580.mp4

2. Layers: show and hide
Made a way to show or hide layers:

https://user-images.githubusercontent.com/50477222/135872526-9cd38553-1806-451c-89eb-12afc6ad9008.mp4

It's not currently the best place, but didn't want to spend time on UX: just click on the movement icon. Anyway, the workflow implemented: imagine a Creator sets background layers and wants a colour layer on `base` responsive stage. When arrives to `m` wants to combine it with an image layer, and creates it. Automatically, that layer is added to the set and automatically hidden in bigger responsive stages (like `l`, `xl`, `xxl`), but on that responsive stages will find it in the library:

https://user-images.githubusercontent.com/50477222/135873648-e613ccf4-a49e-42e8-a97d-a544f954b9fb.mp4

Not sure if is the best UX solution, that's why I share it: to consider. 🤔 

3. Videos and shapes
The difference between `Colour`, `Gradient` and `Image` options with `Video` and `Shape` is on the HTML element. The three firsts work uniquely with CSS, so Style API can deal with the changes. For example, is not a problem to change the image on different responsive stages:

https://user-images.githubusercontent.com/50477222/135975653-c7dfaeb2-9302-4db4-9a38-50ae01ff7b62.mp4

For video and shape on simple background is not that easy:

https://user-images.githubusercontent.com/50477222/135976107-c67b8d46-f21d-4686-9f59-a18deab4eed5.mp4

And with layers is the same: it works good for colour, gradient and image, but can't change the responsive values for video or shape. Need to remark: what can't be changed is the settings that affect the HTML element, like the video url or the SVG code for the shape.

A possible solution for that is to don't allow changes on that settings, but _we should make creators understand_ that behavior. And in that case, if users want to change the video or the shape in some responsive stage, they will need to use `layers background` instead of `simple background` and hide/show depending the necessities.

4. Overwriting CSS with breakpoints:
That happened us on text options (TypographyControl): **some responsive options overwrites `base` responsive stage options**. In the next video case, the `xl` settings 'overwrites' the `base` settings because `xl` is the size that corresponds to my screen, so it shows it:

https://user-images.githubusercontent.com/50477222/135987389-f811d3e4-a9ee-48f8-87bf-de91dabd09c3.mp4

_This is not a bug_, is an expected behavior that may generate a strange UX. The way we fixed on text options has been:
<img width="282" alt="Captura de pantalla 2021-10-05 a las 10 26 34" src="https://user-images.githubusercontent.com/50477222/135987665-fccc956a-d185-44ab-9801-2cbbebc23449.png">

5. Hovers
Hovers is the icing on the cake 🍰 For the moment hover background responsive support hasn't been implemented on this branch as I would like you to test normal features before implementing such a complex job.

-----------

As you see, background with responsive support generates some logical issues on the component and overcomplicates not just the scripts for making it work, also the UX. Before continue working on it and patching all these commented situations (that for sure have a solution) doing complicated things that Creators may find strange, I would like to suggest to **simplify the background workflow**. For example, I think we can fix half of the problem **removing simple background option and leaving just layers background**. At the end, a simple background is the same than layers background with just one layer. With that we simplify 50% the complexity of the component.

Other suggestion is related with the points 3 and 4. This would be an option:

| Colour layer | Video layer |
| ---- | ---- |
| <img width="357" alt="Captura de pantalla 2021-10-05 a las 10 43 29" src="https://user-images.githubusercontent.com/50477222/135990653-9837c018-29f0-4c7b-b458-0f78a05ef66c.png"> | <img width="266" alt="Captura de pantalla 2021-10-05 a las 10 47 28" src="https://user-images.githubusercontent.com/50477222/135995502-2c0eb315-48cb-4154-9628-dbd489479f92.png"> |

It fixes 3, as we can easily make Creators understand (as can be seen in the Video layer example) that some options, like video url, start time or end time, are general options that can't be changed on responsive stages, and others can. In layers like colour, is easy to understand which options are available for responsive (all of them in this case).
Also fixes 4, as users can see the actual breakpoint on the background settings with the red point on it, as it happens in the text settings. I found this was easier to understand for Creators and makes the logicals of the component easier, which transform on easier code to write, maintain and fix in future.

So, resuming:
1. Remove simple background and leave direct access to layers background
2. Set inside layers cards a responsive menu for the options related with the responsive.

What do you think? 🤔

Naaaaiix committed:
Refactor MaxiModal

Naaaaiix committed:
Fix minor bug

Naaaaiix committed:
Merge branch 'modal-refactoring' into issue-1989

Naaaaiix committed:
Fix minor bug
Fix onRemove on MaxiModal

Naaaaiix committed:
Fix minor bug

Naaaaiix committed:
Merge branch 'modal-refactoring' into issue-1989

Naaaaiix committed:
Add icon support to MaxiModal

Naaaaiix committed:
Merge branch 'modal-refactoring' into issue-1989

Naaaaiix committed:
Add responsive support to simple background

Naaaaiix committed:
Merge branch 'master' into issue-1989

Naaaaiix committed:
Update getBackgroundStyles helper
Update getBackgroundStyles tests

Naaaaiix committed:
Fix background layers

Naaaaiix committed:
Fix minor issues
Fix tests

Naaaaiix committed:
Merge branch 'master' into issue-1989

Naaaaiix committed:
Fix keys
Fix tests

Naaaaiix committed:
Fix minor bug

Naaaaiix committed:
Fix Button Maxi background colour
Fix blocks to show background settings in all responsive stages
Fix Arrow background issues
Update tests

Naaaaiix committed:
Merge branch 'master' into issue-1989

Naaaaiix committed:
Fix video layer

Naaaaiix committed:
Merge branch 'master' into issue-1989

Naaaaiix committed:
Remove simple background

Naaaaiix committed:
Create from abstraction ResponsiveTabsControl

Naaaaiix committed:
Add ResponsiveTabsControl to BackgroundControl
Adapts Video and Shape layers

Naaaaiix committed:
Fix minor bugs

Naaaaiix committed:
Abstract BlockBackgroundControl
Fix HoverEffects background support

Naaaaiix committed:
Add default layer to Container Maxi
Clean background attributes on blocks
Fix Button Maxi defaults
Fix Parallax

Naaaaiix committed:
Clean group attributes of BlockBackgroundControl
Adapt arrow to new background
Adapt icon to new background
Fix minor bugs

Naaaaiix committed:
Add 'general' attr support to responsive tabs
Update background helper tests
Fix minor bugs on layers

Naaaaiix committed:
Update integral tests

Naaaaiix committed:
Update parallax integral test
Fix minor bug

Naaaaiix committed:
Fix minor bug

Naaaaiix committed:
Fixed Button Maxi bg blocStyle
Fixed error when removing from image layer
Add background hover responsive support
Add hover tests for background style helper
Fix minor bugs

Naaaaiix committed:
Add buttonless support on LoaderControl
Fix background layers displayer icon
Fix background layers drag
Fix previews on bg layers

Naaaaiix committed:
Fix background on toolbar

Naaaaiix committed:
Fix e2e tests

Naaaaiix committed:
Merge branch 'master' into issue-1989

Naaaaiix committed:
Fix e2e tests

Naaaaiix committed:
Fix integral tests

Naaaaiix committed:
Fix obsolete e2e tests

Naaaaiix committed:
Fix LoaderControl bug

Naaaaiix committed:
Clean test
Fix arrow background colour
Fix minor bugs

Naaaaiix committed:
Fix duplicate layers

Naaaaiix committed:
Fix minor bugs

Naaaaiix committed:
Merge branch 'master' into issue-1989

Naaaaiix committed:
Fix tests

elzadj reviewed (changes requested):
Great work!

Here are some notes, create new issues if needed :+1: :

- I'm getting 
```
react_devtools_backend.js:2526 Warning: Can't perform a React state update on an unmounted component. This is a no-op, but it indicates a memory leak in your application. To fix, cancel all subscriptions and asynchronous tasks in a useEffect cleanup function.
    in Dropdown (created by InsertPoint)
    in InsertPoint (created by CustomGradientBar)
```
when I add additional gradient points to a gradient layer.  

- Can we open the BG layer settings after we add it? Otherwise need to click it every time.
- The hover for BG affects the normal BG - i.e. if I remove a layer from hover, it's removed for the normal too, if I add - it's added too. Also hover has just 2 options for the bg layers - is that intentional?:

https://user-images.githubusercontent.com/2401618/137312912-d245863b-7331-4d2e-957f-bde17abd201a.mp4

- I see (in the console)
```
GET http://wordpress.local/wp-admin/undefined 404 (Not Found)
```
when I add a video or an image layer (see video below).

- The video layer's size is weird on backend (it's ok on frontend):


https://user-images.githubusercontent.com/2401618/137314561-b5994266-88bd-4b0b-baef-8cda98997cb6.mp4

- The button's BG still looks like  the old simple one - is that ok?
- The parallax effect does not work with 
```
front-scripts.js?ver=5.8:42 Uncaught TypeError: Cannot read properties of null (reading 'style')
    at Parallax.animateItem (front-scripts.js?ver=5.8:42)
    at new Parallax (front-scripts.js?ver=5.8:23)
    at front-scripts.js?ver=5.8:332
```

Naaaaiix committed:
Auto-open layers when add
Split background components into blockBackground and background
Abstract background tab from inspectors
Fix minor bugs

Naaaaiix committed:
Fix arrow colour on hover
Fix tests

Naaaaiix committed:
Fix console error

Naaaaiix committed:
Minor fix

elzadj reviewed (changes requested):
Thank you!

> Couldn't reproduce it. Can you share code or steps to do it? thanks  

Can't reproduce it too, but found another one, sorry :weary:  (a Text Block with a parallax bg is crashing after I reload the editor):

https://user-images.githubusercontent.com/2401618/137486050-ffd1cbf2-2264-4793-8c1c-92d0bc9f6493.mp4

> Here's a video about hovering on background:

The hover bg should not be able to remove / add the normal BG layers, this is unexpected behavior:

https://user-images.githubusercontent.com/2401618/137486627-3fd3fe1d-635b-4fa2-b367-f924010d5081.mp4

Can we restrict removing on hover tab somehow :thinking: ?

Naaaaiix committed:
Disable parallax on non necessary blocks

Naaaaiix committed:
Fix adding bg layers on responsive stage
Fix layer mover just on general

Naaaaiix committed:
Fix #1986 (#2221)
